// babel-plugin-flow-runtime/lib/firstPassVisitors.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }();

exports.default = firstPassVisitors;

var _babelTypes = require('babel-types');

var t = _interopRequireWildcard(_babelTypes);

var _getTypeParameters = require('./getTypeParameters');

var _getTypeParameters2 = _interopRequireDefault(_getTypeParameters);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function firstPassVisitors(context) {

  return {
    Identifier: function Identifier(path) {
      var parentPath = path.parentPath;

      if (parentPath.isFlow()) {
        // This identifier might point to a type that has not been resolved yet
        if (parentPath.isTypeAlias() || parentPath.isInterfaceDeclaration()) {
          if (path.key === 'id') {
            return; // this is part of the declaration name
          }
        }
        if (context.hasTDZIssue(path.node.name, path)) {
          context.markBoxed(path.node);
        }
        return;
      } else if (!context.shouldImport) {
        return;
      }
      if (path.key === 'property' && parentPath.isMemberExpression() && parentPath.node.computed) {
        return;
      }
      var name = path.node.name;

      if (name === context.libraryId) {
        context.libraryId = path.scope.generateUid(context.libraryId);
      }
    },
    TypeAlias: function TypeAlias(path) {
      context.defineTypeAlias(path.node.id.name, path);
    },
    InterfaceDeclaration: function InterfaceDeclaration(path) {
      context.defineTypeAlias(path.node.id.name, path);
    },
    ImportDeclaration: function ImportDeclaration(path) {
      var source = path.get('source').node.value;

      var isReact = path.node.importKind === 'value' && (source === 'react' || source === 'preact');

      var isFlowRuntime = path.node.importKind === 'value' && source === 'flow-runtime';

      if (isReact) {
        path.parentPath.scope.setData('importsReact', true);
      }

      path.get('specifiers').forEach(function (specifier) {
        var local = specifier.get('local');
        var name = local.node.name;

        if (path.node.importKind === 'type') {
          context.defineImportedType(name, specifier);
        } else {
          context.defineValue(name, path);
          if (isReact) {
            if (specifier.isImportDefaultSpecifier()) {
              path.parentPath.scope.setData('reactLib', name);
            } else if (specifier.isImportNamespaceSpecifier()) {
              path.parentPath.scope.setData('reactLib', name);
            } else if (specifier.node.imported.name === 'Component') {
              path.parentPath.scope.setData('reactComponentClass', name);
            } else if (specifier.node.imported.name === 'PureComponent') {
              path.parentPath.scope.setData('reactPureComponentClass', name);
            }
          } else if (isFlowRuntime && (specifier.isImportDefaultSpecifier() || specifier.isImportNamespaceSpecifier())) {
            context.shouldImport = false;
            context.libraryId = name;
          }
        }
      });
    },
    VariableDeclarator: function VariableDeclarator(path) {
      var name = path.node.id.name;

      context.defineValue(name, path);
    },
    Function: function Function(path) {
      if (path.isFunctionDeclaration() && path.has('id')) {
        var name = path.node.id.name;

        context.defineValue(name, path.parentPath);
      }
      var params = path.get('params').filter(hasTypeAnnotation);
      var typeParameters = (0, _getTypeParameters2.default)(path);
      if (path.has('returnType') || params.length || typeParameters.length) {
        var body = path.get('body');

        if (!body.isBlockStatement()) {
          // Expand arrow function expressions
          body.replaceWith(t.blockStatement([t.returnStatement(body.node)]));
          body = path.get('body');
        }

        typeParameters.forEach(function (item) {
          var name = item.node.name;

          context.defineTypeParameter(name, item);
        });
        params.forEach(function (param) {
          context.defineTypeAlias(param.node.name, param);
        });
      }
    },
    Class: function Class(path) {
      var className = 'AnonymousClass';
      if (path.isClassDeclaration() && path.has('id')) {
        var name = path.node.id.name;

        className = name;
        context.defineValue(name, path.parentPath);
      }
      context.setClassData(path, 'currentClassName', className);
      var typeParameters = (0, _getTypeParameters2.default)(path);
      typeParameters.forEach(function (item) {
        var name = item.node.name;

        context.defineClassTypeParameter(name, item);
      });
      if (typeParameters.length > 0 || path.has('superTypeParameters')) {
        ensureConstructor(path);
        context.setClassData(path, 'typeParametersUid', path.parentPath.scope.generateUid('_typeParameters'));
      }

      if (typeParameters.length > 0) {
        context.setClassData(path, 'typeParametersSymbolUid', path.parentPath.scope.generateUid(className + 'TypeParametersSymbol'));
      } else {
        context.setClassData(path, 'typeParametersSymbolUid', '');
      }
    }
  };
}

/**
 * Determine whether the given node path has a type annotation or not.
 */
function hasTypeAnnotation(path) {
  return path.node && path.node.typeAnnotation ? true : false;
}

/**
 * Ensure that the given class contains a constructor.
 */
function ensureConstructor(path) {
  var lastProperty = void 0;

  var _path$get$filter = path.get('body.body').filter(function (item) {
    if (item.isClassProperty()) {
      lastProperty = item;
      return false;
    }
    return item.node.kind === 'constructor';
  }),
      _path$get$filter2 = _slicedToArray(_path$get$filter, 1),
      existing = _path$get$filter2[0];

  if (existing) {
    return existing;
  }
  var constructorNode = void 0;
  if (path.has('superClass')) {
    var args = t.identifier('args');
    constructorNode = t.classMethod('constructor', t.identifier('constructor'), [t.restElement(args)], t.blockStatement([t.expressionStatement(t.callExpression(t.super(), [t.spreadElement(args)]))]));
  } else {
    constructorNode = t.classMethod('constructor', t.identifier('constructor'), [], t.blockStatement([]));
  }

  if (lastProperty) {
    lastProperty.insertAfter(constructorNode);
  } else {
    path.get('body').unshiftContainer('body', constructorNode);
  }
}
